lib/json-schema/util/uri.rb

```diff
@@ -11,6 +16,11 @@ def self.normalized_uri(uri)
         end
         uri
       end
+
+      def self.normalized_uri(uri)
+        instance.cache ||= {}
+        instance.cache[uri] ||= instance.normalized_uri(uri)
```

Comment:
@iainbeeston right, the only effect of the race condition is a cache miss, and that's OK.